### PR TITLE
CMake flags cleanup

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This PR cleans up the cflags and cxxflags used for building arangod and the client tools.

It removes the duplicate usage of the same flags inside each compile command, which previously happened because we set CMake options like as follows:
```
    CMAKE_CXX_FLAGS=-Wall -Wextra ...
    CMAKE_CXX_FLAGS_DEBUG=${CMAKE_CXX_FLAGS} ... (more flags)
    CMAKE_CXX_FLAGS_RELEASE=${CMAKE_CXX_FLAGS} ... (more flags)
    CMAKE_CXX_FLAGS_RELWITHDEBINFO=${CMAKE_CXX_FLAGS} ... (more flags)
```
As can be seen, the CMAKE_CXX_FLAGS were copied into the build type-specific flags CMAKE_CXX_FLAGS_DEBUG, CMAKE_CXX_FLAGS_RELEASE, CMAKE_CXX_FLAGS_RELWITHDEBINFO etc.
However, CMake automatically concatenates the base flags in CMAKE_CXX_FLAGS with the build-type specific flags in CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE}, so we were effectively seeing the base flags multiple times on each
compiler invocation.

In addition, this PR introduces a new CMake option `USE_MINIMAL_DEBUGINFO`. 
This option is turned off by default. If turned on, the created binaries will contain only a minimum amount of debug symbols, reducing the size of the executables. If turned off (which is the default), the binaries will contain full debug information, which will make them bigger in size unless the debug information is later stripped again.

Finally, the PR turns on the `-Wsuggest-overrides` flag for g++, which will warn about missing `override` specifiers. Note that this will make future build fail when compiling with `USE_FAIL_ON_WARNINGS=On`.

- [ ] :hankey: Bugfix 
- [x] :pizza: New feature 
- [x] :hammer: Refactoring 
- [x] :book: CHANGELOG entry made
- [x] :muscle: The behavior in this PR was *manually tested*
- [x] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] No backports required

### Testing & Verification

Compilation must still succeed, even with new warnings turned on.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12438/

### CODE REVIEW

**Comment by jsteemann** [comment]
Tests blue

**Review by fceller — APPROVED** [review]
LGTM

File: CMakeLists.txt [hunk 1]
```diff
@@ -891,110 +893,105 @@ endif ()
...
-  set(CMAKE_C_FLAGS                "-g"                             CACHE INTERNAL "default C compiler flags")
-  set(CMAKE_C_FLAGS_DEBUG          "-O0 -g -D_DEBUG=1"              CACHE INTERNAL "C debug flags")
-  set(CMAKE_C_FLAGS_MINSIZEREL     "-Os"                            CACHE INTERNAL "C minimal size flags")
-  set(CMAKE_C_FLAGS_RELEASE        "-O3 -fomit-frame-pointer"       CACHE INTERNAL "C release flags")
-  set(CMAKE_C_FLAGS_RELWITHDEBINFO "-O3 -g -fno-omit-frame-pointer" CACHE INTERNAL "C release with debug info flags")
+  set(CMAKE_C_FLAGS                "/MTd"                                              CACHE INTERNAL "default C++ compiler flags")
+  set(CMAKE_C_FLAGS_DEBUG          "/D _DEBUG /MTd /Zi /Ob0 /Od /RTC1 /bigobj"         CACHE INTERNAL "C++ debug flags")
+  set(CMAKE_C_FLAGS_MINSIZEREL     "/MT /O1 /Ob1"                                      CACHE INTERNAL "C++ minimal size flags")
+  set(CMAKE_C_FLAGS_RELEASE        "/MT /O2 /Ob2"                                      CACHE INTERNAL "C++ release flags")
+  set(CMAKE_C_FLAGS_RELWITHDEBINFO "/MT /Zi /O2 /Ob1"                                  CACHE INTERNAL "C++ release with debug info flags")
```
Comment (KVS85) [inline CMakeLists.txt:909]: Why `C++` comment for `CMAKE_C_FLAGS`?

**Review by KVS85 — COMMENTED** [review]
LGTM.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12497/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12504/

File: CMakeLists.txt [hunk 2]
```diff
@@ -891,110 +893,105 @@ endif ()
...
-  set(CMAKE_CXX_FLAGS                "-g -Wnon-virtual-dtor"          CACHE INTERNAL "default C++ compiler flags")
-  set(CMAKE_CXX_FLAGS_DEBUG          "-O0 -g -D_DEBUG=1"              CACHE INTERNAL "C++ debug flags")
-  set(CMAKE_CXX_FLAGS_MINSIZEREL     "-Os"                            CACHE INTERNAL "C++ minimal size flags")
-  set(CMAKE_CXX_FLAGS_RELEASE        "-O3 -fomit-frame-pointer"       CACHE INTERNAL "C++ release flags")
-  set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g -fno-omit-frame-pointer" CACHE INTERNAL "C++ release with debug info flags")
-
-elseif (MSVC)
-  if (VERBOSE)
-    message(STATUS "Compiler type MSVC: ${CMAKE_CXX_COMPILER}")
+  # all non-MSVC compilers go here...
```
Comment (dothebart) [inline CMakeLists.txt:926]: this comment is unneeded.

File: CMakeLists.txt [hunk 3]
```diff
@@ -891,110 +893,105 @@ endif ()
...
+  if (CMAKE_COMPILER_IS_GNUCC)
+    # g++
```
Comment (dothebart) [inline CMakeLists.txt:928]: stating the obvious - remove.

File: CMakeLists.txt [hunk 4]
```diff
@@ -891,110 +893,105 @@ endif ()
...
+    message(STATUS "Compiler type GNU: ${CMAKE_CXX_COMPILER}")
+    set(BASE_FLAGS "-Wall -Wextra -Wno-unused-parameter -Wno-deprecated-declarations ${BASE_FLAGS}")
+    set(EXTRA_CXX_FLAGS "-Wsuggest-override -Wnon-virtual-dtor")
+  elseif (CMAKE_COMPILER_IS_CLANG)
+    # clang
```
Comment (dothebart) [inline CMakeLists.txt:933]: obvious again - remove.

**Review by dothebart — APPROVED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (jsteemann) [inline CMakeLists.txt:926]:
```suggestion
```

File: CMakeLists.txt (hunk 4 above)
Comment (jsteemann) [inline CMakeLists.txt:933]:
```suggestion
```

File: CMakeLists.txt (hunk 3 above)
Comment (jsteemann) [inline CMakeLists.txt:928]:
```suggestion
```

**Pull request merged**